src/Http/Http/src/StreamPipeReader.cs

```diff
@@ -248,6 +248,13 @@ public override async ValueTask<ReadResult> ReadAsync(CancellationToken cancella
 
                     _readTail.End += length;
                     _bufferedBytes += length;
+
+                    if (length == 0)
+                    {
+                        // If Stream.ReadAsync returns empty, it means reading is done.
+                        // Complete the PipeReader
+                        Complete();
```

Comment:
This was a bug that surfaced due to StreamPipe interop tests. If a stream returns 0 bytes, I think it means the PipeReader is done, so we can call Complete().

Reply:
> PipeReader.Complete() should only be called by the reader once the read loop is completed.

So only call Complete internally on Dispose. All other calls to Complete will happen externally?

Reply:
That’s right. The caller and dispose call complete.

Reply:
> Additionally, if PipeReader.AdvanceTo() is called after PipeReader.Complete(), which would likely happen with the completion of every Stream with this change left as is, PipeReader.AdvanceTo() is supposed to throw.

What about ReadAsync/TryRead?

Reply:
Same.